### DISCUSSION TITLE
Make Interface Discovery easier

### DISCUSSION THREAD

**Discussion by MiniDigger** [directus/directus#4044] [Feature Requests]
When creating collections/fields, there are two kind of users. The more technical ones and the more UX focused ones.

The first ones mostly care about how they data is saved in the DB first, and then care about how to display is nicely in the UI for editing. For the other group, they first care about how to insert the later, and deal with actually retrieving/the storage later.

Currently, directus caters to the first group, by making you select the datatype first and then the interface. This makes it very hard for the second crowd to find the interface they want, often the only solution is to manually go thru all types and find the interface you want (if that even exists).

recent quote by someone on discord: "for me the fact that a tags input is stored as a CSV field in the database is an implentation detail. [...] When building my model I thought "I want this field to be a tags input"."

a solution to this would be to unify the interface and type selection step. I for example could see a accordion like UI, where the headers are the datatypes and collapsed below that a list of interfaces that support that type. in the next steps you would then have the current DB settings and then the interface settings. 
Somebody might a better since for UX might have a better idea to solve this problem, but I think this illustrates what I am proposing pretty nice.

**Comment by benhaynes** [comment]
This is a great idea. We had it "interface first" in v8, and then wanted more of a data-first approach in v9, so we flipped it. Though it can be hard to know/find the interface you're looking for based on the true datatype. I like this idea of presenting _both_, as in an accordion... or we could have two options at the beginning (eg: **standard** and **advanced**).

I'll see what I can come up with on the design side to clean this up. We also wanted to optimize the relational setup flow, so maybe it's a good time to re-think this whole process. I don't think it needs a huge code overhaul (don't worry, @rijkvanzanten), but some UX/UI polish would help!

**Reply by benhaynes** [reply]
OK, here is a first draft of the new field setup flow. I'm thinking everything stays the same, but we move the "Choose Field Type" from the dropdown it is now, to a dedicated drawer pane. 

1. Browse the interfaces by _very_ high-level categories (Text & Numbers, Selectable, Presentation, Other)
2. Click the interface you want
3. If there's more than one supported datatype, clicking reveals a dropdown to choose which you want.
4. Advance on to the same flow as now, but with interface/type locked.

![Settings-Data-Model-Field-Create-1A](https://user-images.githubusercontent.com/522079/108118252-11a90300-706c-11eb-9f34-475a622fc8c8.png)

Thoughts? Hopes? Wishes? Dreams?

**Reply by rijkvanzanten** [reply]
Love it! The only thing I'm worrying about is the active state + select a type to continue. I'm not sure if it's clear enough what you're supposed to do at that point 🤔

**Reply by benhaynes** [reply]
Yeah, we could split that into a second step, but that seems like overkill. The MAIN click event could just open the context dropdown to choose a type, then it can't be missed. eg:

```
- Save as JSON
- Save as CSV
- Save as String
```

**Reply by philleepflorence** [reply]
From documentation discovery found these **content types** that may work for **end** and **tech** users:

1. String - Single Line - title, names, email address, password
2. Text - Multiple Lines - descriptions, synopsis, text paragraphs
3. Rich Text - WYSIWYG - Formatted text
4. Number - Integer, Floats, Decimals - Prices or anything that is a number
5. Boolean - Yes or No, True or False, 1 or 0
6. Date - DateTime, Timestamp - Event Date, Hours
7. Media - Files - Images, videos, PDFs, and other files
8. Relation - Link to another content
9. JSON - Data in JSON format
10. Enumeration - CSV - List of choices
11. Group - Repeater - Combination of other fields

Not sure how granular we want to go but wanted to paste the summary here.

**Reply by benhaynes** [reply]
Those are from our Directus type superset, which are less granular than vendor lists, but two granular for this interface, I think.

Each of the category headings takes a certain amount of vertical room, and doesn't necessarily add a lot of initial context.

Also, by having only three very high-level categories we're able to put off the type decision until after you choose an interface. That is the purpose of the type drop down when clicking the interface.

So as of now, I think this simple categorization might be ideal.

**Reply by philleepflorence** [reply]
Fair enough...mostly wanted to share what I saw as common across various vendors.

BTW, where is the Directus type superset?
It would be interesting that the list is almost identical.

**Reply by MiniDigger** [reply]
thats basically exactly what I had in mind (just, looking actually good :P)

one thing I noticed that my poor custom interfaces don't have fancy pictures, would that be something you can define in an extension maybe?

**Reply by benhaynes** [reply]
@philleepflorence — in the docs: https://docs.directus.io/concepts/types/#directus-data-type-superset

@MiniDigger — Yeah! I was thinking that custom extensions could use an icon (slightly stylized). Let me mock something up for that!

**Reply by benhaynes** [reply]
Here are a few options for how we could handle custom extensions and keep them visually similar. I imagine them setting a category and icon for organization.

![Screen Shot 2021-02-17 at 12 51 58 PM](https://user-images.githubusercontent.com/522079/108246041-fe0aa480-711e-11eb-90ff-d15bc600476f.png)

![Screen Shot 2021-02-17 at 12 52 11 PM](https://user-images.githubusercontent.com/522079/108246027-f9de8700-711e-11eb-8e6d-623df3a8bca4.png)

**Reply by philleepflorence** [reply]
@benhaynes - Thanks!

**Comment by brpaz** [comment]
Hello, I am the one who put that question on Discord. ;)

I wanted to create a field that was a "tags input" and took me a while to find that I needed to select the CSV data type first.

In my mind, I was expecting to find it maybe as an interface for the text input like the dropdown, for example. I would never associate with the word CSV at first.
As I said on Discord the fact the tags are stored as CSV field by Directus is an implementation detail that I shouldn´t need to know when creating my Data model. (It could store it as a one to many in a separated tags table for example instead). 

And I am a developer so I think there is some space for improvement here. ;)

**Comment by lukas-schaetzle** [comment]
Just wanted to point out the same thing.
@benhaynes I absolutely love the mockup +1 And I would also include the relational fields in there (many-2-one, many-2-many, ...). Maybe the "Dynamic" option in the mockup is meant for that?

**Reply by rijkvanzanten** [reply]
> And I would also include the relational fields in there (many-2-one, many-2-many, ...). Maybe the "Dynamic" option in the mockup is meant for that?

We were kinda thinking we'd "include" those under the visual representation 🤔 

For example, the dropdown can be a "simple string" _or_ a many-to-one, the repeater can be "inline JSON" _or_ a one-to-many etc

**Reply by lukas-schaetzle** [reply]
Yeah, that makes sense. Sounds good.

**Reply by benhaynes** [reply]
Exactly. That said, maybe it would be nice to show how many types are supported by each interface style without having to click on it. Could be icons for static vs relational data support, or a count of supported types... but perhaps a very clear hover state will suffice.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
